docs/05-rfcs/2023-01-21-wing-plugins.md

```diff
@@ -0,0 +1,115 @@
+
+- An organization requires its cloud resources to follow strict security requirements to maintain their HIPPA compliance. These rules include encryption at rest, versioning, replication, isolation, and more. These rules can be encapsulated in a plugin that can be applied to all resources.
+
+- An application team is provided a VPC in their account that was provisioned by the platform engineering team. It is required that all the resources in the application are deployed into this VPC.
+
+- Security policies in an organization require the use of customer managed keys for all resources that support encryption at rest. 
+
+- To better manage and understand costs, an organization requires that all resources are tagged with a certain set of tags.
+
+- An organization's security team has decided that all IAM policies should not contain Allow "*" actions. A plugin can be written to scan the synthesized config and throw an error if a policy statement is found, and where it is found.
+
+- A development team is importing an existing stateful resource into their wing code, however the access to the resource should only ever be readonly. The devops team decides to write a plugin to ensure that no resource in the wing code is ever granted a policy that would allow write access to the imported stateful resource.
+
+## Examples
+These examples are meant to be illustrative of the plugin system, and not necessarily the best way to implement the use cases.
+
+Using plugin to apply existing cdk aspects
+```ts
+// some existing aspect
+class HelloAspect implements IAspect {
+  visit(node: IConstruct) {
+    console.log(`Hello, ${node.node.path}`);
+  }
+}
+
+// my-plugin.js
+class MyPlugin extends WingPlugin {
+  preSynthesis(app) {
+    Aspects.of(app).add(new HelloAspect());
+  }
+
+  postSynthesis(config) {
+    // no-op
+  }
+}
+```
+
+Inspect synthesized template for wildcards
+```ts
+// my-plugin.js
+class WildcardPolicyScanner extends WingPlugin {
+  preSynthesis(app) {
+    // no-op
+  }
+
+  postSynthesis(config) {
+    const resources = config.resource;
+    if (resources.aws_iam_role_policy) {
+      const policyKeys = Object.keys(resources.aws_iam_role_policy);
+      
+      for (const key of policyKeys) {
+        const policy = resources.aws_iam_role_policy[key].policy;
+        const statements = JSON.parse(policy).Statement;
+        
+        for (const statement of statements) {
+          const actions = statement.Action as string[];
+          
+          if (actions.includes("*") && statement.Effect === "Allow") {
+            throw new Error(`Found a wildcard allow action in policy: ${key}`);
+          }
+        }
+      }
+    }
+  }
+}
```

Comment:
Can you also describe how you are going to "plug" the plugin?  Is this a compiler flag?